File: crates/examples/agent_rpc_demo.py
#!/usr/bin/env python3
"""
Agent RPC Server Demo — Expose an agent as a callable server, call from another agent

Demonstrates:
1. Creating Agent A with MCP tools and exposing it as a callable server
2. Creating Agent B that calls Agent A via RPC (tool/list, tool/call, llm/run)
3. Agent-to-agent communication over the bus

Architecture:
    Agent B (RPC client) ──bus──> Agent A (callable server + MCP tools)
                                       │
                                       └──> npx mcp-hello-world (echo, add, debug)

Usage:
    export OPENAI_API_KEY="sk-..."
    export LLM_BASE_URL="https://integrate.api.nvidia.com/v1"
    export LLM_MODEL="nvidia/meta/llama-3.1-8b-instruct"
    python3 crates/examples/agent_rpc_demo.py
"""

import asyncio
import json
import os

from pybos import Agent, AgentConfig, Bus, BusConfig, ConfigLoader

loader = ConfigLoader()
loader.discover()
_config = loader.load_sync()
_global = _config.get("global_model", {})

API_KEY = os.environ.get("OPENAI_API_KEY") or _global.get("api_key", "")
BASE_URL = os.environ.get("LLM_BASE_URL") or _global.get("base_url", "https://integrate.api.nvidia.com/v1")
MODEL = os.environ.get("LLM_MODEL") or _global.get("model", "nvidia/meta/llama-3.1-8b-instruct")


async def demo_agent_rpc():
    print("═" * 60)
    print("  Demo — Agent A as callable server, Agent B as RPC client")
    print("═" * 60)

    bus = await Bus.create(BusConfig())

    # ── Agent A: has MCP tools, exposed as callable server ──
    print("\n  ── Setting up Agent A (server side) ──")

    config_a = AgentConfig(
        name="agent-a",
        model=MODEL,
        base_url=BASE_URL,
        api_key=API_KEY,
        system_prompt="You are a helpful assistant with math and greeting tools.",
        temperature=0.7,
        timeout_secs=120,
    )
    agent_a = await Agent.create(config_a, bus)
    print("  🤖 Agent A created")

    await agent_a.add_mcp_server("hello", "npx", ["-y", "mcp-hello-world@latest"])
    print("  🔌 Agent A connected to MCP hello-world server")

    mcp_tools = await agent_a.list_mcp_tools()
    print(f"  🔧 Agent A MCP tools: {[t.get('name') for t in mcp_tools]}")

    server = await agent_a.as_callable_server("zenoh/agent-a", bus)
    print(f"  📡 Agent A callable server started at: {server.endpoint()}")
    print(f"     is_started: {server.is_started()}")

    # ── Agent B: RPC client calling Agent A ──
    print("\n  ── Setting up Agent B (client side) ──")

    config_b = AgentConfig(
        name="agent-b",
        model=MODEL,
        base_url=BASE_URL,
        api_key=API_KEY,
        system_prompt="You are a coordinator that calls other agents via RPC.",
        temperature=0.7,
        timeout_secs=120,
    )
    agent_b = await Agent.create(config_b, bus)
    print("  🤖 Agent B created")

    rpc = await agent_b.rpc_client("zenoh/agent-a", bus)
    print(f"  🔗 Agent B RPC client connected to: {rpc.endpoint()}")

    # ── RPC: tool/list ──
    print("\n  ── RPC: tool/list ──")
    tools = await rpc.list()
    print(f"  📋 Agent A's tools: {tools}")

    # ── RPC: tool/call ──
    print("\n  ── RPC: tool/call ──")
    result = await rpc.call("hello/add", json.dumps({"a": 7, "b": 8}))
    print(f"  📤 call(hello/add, {{a:7, b:8}})")
    print(f"  📥 Result: {result}")

    # ── RPC: llm/run ──
    print("\n  ── RPC: llm/run ──")
    print(f"  📤 llm_run('What is 3 + 5?')")
    try:
        reply = await rpc.llm_run("What is 3 + 5? Use the hello/add tool.")
        text = reply.get("text", "") if isinstance(reply, dict) else str(reply)
        print(f"  📥 Agent A reply: {text[:300]}")
    except Exception as e:
        print(f"  ⚠️  llm_run timed out or failed (LLM takes >10s over bus RPC): {e}")
        print(f"  ℹ️  Use tool/call for fast tool invocations; llm/run needs longer timeout")

    print("\n  ✅ Agent RPC demo done\n")


async def main():
    print("\n" + "🔗" * 30)
    print("  BrainOS — Agent RPC Server Demo")
    print("🔗" * 30 + "\n")

    if not API_KEY:
        print("  ⚠️  OPENAI_API_KEY not set — LLM calls will fail")
        print("  Set: export OPENAI_API_KEY=sk-...\n")

    await demo_agent_rpc()

    print("═" * 60)
    print("  ✅ All Agent RPC demos completed!")
    print("═" * 60 + "\n")


if __name__ == "__main__":
    asyncio.run(main())

File: crates/jsbos/examples/agent_rpc_demo.js
#!/usr/bin/env node
/**
 * Agent RPC Server Demo — Expose an agent as a callable server, call from another agent
 *
 * Note: This demo shows MCP tool integration. Full RPC server functionality
 * requires agent creation with bus session (not yet exposed in JS bindings).
 *
 * Demonstrates:
 * 1. Creating Agent with MCP tools
 * 2. Listing and using MCP tools via agent
 *
 * Usage:
 * export OPENAI_API_KEY="sk-..."
 * export LLM_BASE_URL="https://integrate.api.nvidia.com/v1"
 * export LLM_MODEL="nvidia/meta/llama-3.1-8b-instruct"
 * node crates/jsbos/examples/agent_rpc_demo.js
 */

import { Bus, Agent, ConfigLoader, version, initTracing } from '../index.js'

initTracing()

const loader = new ConfigLoader()
loader.discover()
const _config = JSON.parse(loader.loadSync())
const _global = _config.global_model || {}

const API_KEY = process.env.OPENAI_API_KEY || _global.api_key || ''
const BASE_URL = process.env.LLM_BASE_URL || _global.base_url || 'https://integrate.api.nvidia.com/v1'
const MODEL = process.env.LLM_MODEL || _global.model || 'nvidia/meta/llama-3.1-8b-instruct'

async function demoAgentMcp() {
  console.log('═'.repeat(60))
  console.log(' Demo — Agent with MCP tools')
  console.log('═'.repeat(60))

  const bus = await Bus.create()
  console.log(' 🚌 Bus created')

  console.log('\n ── Setting up Agent with MCP tools ──')

  const configA = {
    name: 'agent-with-mcp',
    model: MODEL,
    baseUrl: BASE_URL,
    apiKey: API_KEY,
    systemPrompt: 'You are a helpful assistant with math and greeting tools.',
    temperature: 0.7,
    timeoutSecs: 120,
  }
  const agent = await Agent.create(configA, bus)
  console.log(' 🤖 Agent created')

  await agent.addMcpServer('hello', 'npx', ['-y', 'mcp-hello-world@latest'])
  console.log(' 🔌 Agent connected to MCP hello-world server')

  await new Promise(resolve => setTimeout(resolve, 2000))

  const mcpTools = await agent.listMcpTools()
  console.log(` 🔧 MCP tools available: ${mcpTools.map(t => t.name).join(', ')}`)

  console.log('\n ── Test: Agent react() with MCP tool ──')
  console.log(" 📤 User: 'Use the hello/add tool to calculate 7 + 8'")
  try {
    const result = await agent.react('Use the hello/add tool to calculate 7 + 8')
    console.log(` 📥 Agent: ${result.slice(0, 200)}...`)
  } catch (e) {
    console.log(` ⚠️ React failed: ${e.message}`)
  }

  console.log('\n ✅ Agent MCP demo done\n')
}

async function main() {
  console.log('\n' + '🔗'.repeat(30))
  console.log(' BrainOS — Agent MCP Demo')
  console.log('🔗'.repeat(30) + '\n')

  if (!API_KEY) {
    console.log(' ⚠️ OPENAI_API_KEY not set — LLM calls will fail')
    console.log(' Set: export OPENAI_API_KEY=sk-...\n')
  }

  await demoAgentMcp()

  console.log('═'.repeat(60))
  console.log(' ✅ All Agent demos completed!')
  console.log('═'.repeat(60) + '\n')
}

main()